### PR TITLE
Static + Dynamic PowerSync Builds

### PR DESCRIPTION
**Pull request by stevensJourney**
# Overview

This supersedes the following PRs:
-  https://github.com/powersync-ja/wa-sqlite/pull/16
- https://github.com/powersync-ja/wa-sqlite/pull/13

## Update

This PR updates our fork with the latest changes from upstream. Our PowerSync core Git submodule had build issues when just pulling in the latest changes. Build issues have been resolved by removing the use of Git submodules in this repository.

### Noteable updates

- Upstream now supports table change notifications. Our old implementation has been removed.
- Upstream has removed `prepare_v2` from `SQLiteAPI`. Our SDK now has to use a different method for `batchExecute`.
- WA SQLite is now on version `1.0.4`. This PR includes a `major` version bump to `@journeyapps/wa-sqlite@1.0.0`

## PowerSync Core

The PowerSync core is no longer provided from a Git submodule. The `libpowersync-wasm.a` binary is now downloaded from the `powersync-sqlite-core` releases during compilation. The PowerSync core is statically linked in the `dist/wa-sqlite.wasm` and `dist/wa-sqlite-async.wasm` distributables.

The PowerSync specific source has been separated from the SQLite source.

The PowerSync extension can be loaded in the static build with 

```js
const { default: moduleFactory } = await import('@journeyapps/dist/wa-sqlite.mjs');
  const module = await moduleFactory();

  module.ccall('setup_powersync', 'int', []);
  const sqlite3 = SQLite.Factory(module);
```

## Dynamic extension loading

See [here](https://github.com/powersync-ja/wa-sqlite/pull/16)  and [here](https://github.com/powersync-ja/powersync-sqlite-core/pull/35) for reference.

We now provide `dist/wa-sqlite-dynamic-main.wasm` and `dist/wa-sqlite-async-dynamic-main.wasm` distributables which support dynamic extension loading. These distributables are compiled as [Emscripten main modules](https://emscripten.org/docs/compiling/Dynamic-Linking.html). 

The PowerSync side module WASM files are configured to be downloaded in a `postinstall` script. 

The PowerSync extension can be loaded with:

```js
const { default: moduleFactory } = await import('@journeyapps/dist/wa-sqlite-dynamic-main.mjs');
  const module = await moduleFactory({
    // locateFile(path) {
    //   // (OPTIONAL) locateFile can be specified to change WASM module resolution.
    //   // The returned url is passed to fetch, and can be a relative path
    //   // or absolute URL.
    //   // This applies to both the main wasm file, and the extension.
    //   return `../dist/${path}`;
    // },
  });

  const sqlite3 = SQLite.Factory(module);
  // Load the extension library into this scope
  let extScope = {};
  await module.loadDynamicLibrary(
     'libpowersync.wasm',
      { loadAsync: true },
      extScope
    );

    // This calls sqlite3_auto_extension(sqlite3_powersync_init).
    // For generic extensions, we'd need to call sqlite3_auto_extension directly.
    extScope.powersync_init_static();
```

Runtime errors could occur if the main module does not export all the required symbols which side modules require. We currently don't have a full list of all the common/required symbols for our PowerSync side module. This list could also change if other SQLite extensions are loaded. In order to ensure stability, the main module is currently compiled with the option `MAIN_MODULE=1`.  This has the unfortunate affect of a large file-size (3.7mb).

### CODE REVIEW

**Review by rkistner — DISMISSED** [review]

File: powersync-static/setup_powersync.c [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#include <sqlite3.h>
+
+extern int sqlite3_powersync_init(sqlite3 *db, char **pzErrMsg,
+                                  const sqlite3_api_routines *pApi);
+
+int setup_powersync()
+{
+  return sqlite3_auto_extension((void (*)(void)) & sqlite3_powersync_init);
+}
```
Comment (rkistner) [inline powersync-static/setup_powersync.c:9]: You can probably skip the custom C code completely, and just use the existing `powersync_init_static` function (does exactly the same as this one).

Comment (stevensJourney) [reply powersync-static/setup_powersync.c]: That is nicer. I've removed this function and file now. `powersync_init_static` is called directly in the demo and in our SDK.

**Review by rkistner — APPROVED** [review]

**Comment by markstos** [comment]
Why use a `postinstall` hook instead a regular dependency? 

Postinstall hooks bring up concerns about security and reproducibility because they are dynamically running additional code at install time.

**Comment by rkistner** [comment]
@markstos This is an ancient PR; it's best to create a new issue.

**Comment by simolus3** [comment]
I see we don't have issues on this fork - maybe we should enable them given that this library is also used in a few places outside of the SDK now?

Either way, the dynamic builds referenced by the download script don't exist anymore and we should be able to remove the postinstall hook. I can take a look at that soon.

**Comment by rkistner** [comment]
Good point - I enabled issues now

**Pull request merged**